### PR TITLE
Add PaymentRequest and Payment messages

### PR DESCRIPTION
**Pull request by x1m3**

### CODE REVIEW

File: protocol/payment.go [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+		}
+		col = append(col, single)
+	}
+	*p = col
+	return nil
+}
+
+// EthereumEip712Signature2021 represents the Ethereum EIP712 signature.
+type EthereumEip712Signature2021 struct {
+	Type               string     `json:"type"`
```
Comment (vmidyllic) [inline protocol/payment.go:197]: verifiable.ProofType

**Comment by vmidyllic** [comment]
do not merge til erc20 payment spec, please

cc: @volodymyr-basiuk

File: protocol/payment.go [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/iden3/iden3comm/v2"
+)
+
+const (
+	// PaymentRequestMessageType is a Iden3PaymentMessage payment type
+	PaymentRequestMessageType iden3comm.ProtocolMessage = iden3comm.DidCommProtocol + "credentials/0.1/payment-request"
```
[the remark this answers is not in the document]
Comment (x1m3) [reply protocol/payment.go]: Also changed PaymentMessageType to same Iden3Protocol domain.

File: protocol/payment.go [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+	if err := json.Unmarshal(data, &strCol); err == nil {
+		p.strCol = strCol
+		return nil
+	}
+	if err := json.Unmarshal(data, &itemCol); err == nil {
+		p.itemCol = itemCol
+		return nil
+	}
+	return errors.Errorf("failed to unmarshal PaymentContext: %s", string(data))
+}
```
[the remark this answers is not in the document]
Comment (x1m3) [reply protocol/payment.go:381]: I agree. I like your proposal. Let me change it.

File: protocol/payment.go [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+
+// Iden3PaymentRailsERC20RequestV1 represents the Iden3PaymentRailsERC20RequestV1 payment request data.
+type Iden3PaymentRailsERC20RequestV1 struct {
+	Nonce          string                         `json:"nonce"`
+	Type           string                         `json:"type"`
+	Context        PaymentContext                 `json:"@context"`
+	Recipient      string                         `json:"recipient"`
+	Amount         string                         `json:"amount"` // Not negative number
+	ExpirationDate string                         `json:"expirationDate"`
+	Proof          EthereumEip712Signature2021Col `json:"proof"`
```
Comment (vmidyllic) [inline protocol/payment.go:236]: I have one doubt regarding that. From the protocol persperctive we can have different proof types here. So we can design structure not to break interfaces later.

What do you think @x1m3 @volodymyr-basiuk ?

Comment (vmidyllic) [reply protocol/payment.go:236]: also it can be an object, not always an array

Comment (x1m3) [reply protocol/payment.go:236]: Yes. I agree.  We need to handle objects and arrays and change Proof structure in a way that can be extended in the future.

Comment (x1m3) [reply protocol/payment.go]: Fixed. I changed it to class that can be extended in the future with other proof types.

File: protocol/payment.go [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+	Data        PaymentRequestInfoData          `json:"data"`
+}
+
+// PaymentRequestInfoData is a union type for field Data in PaymentRequestInfo.
+// Only one of the fields can be set at a time.
+type PaymentRequestInfoData struct {
+	dataType string
+	crypto   []Iden3PaymentRequestCryptoV1
+	rails    []Iden3PaymentRailsRequestV1
+	railsERC []Iden3PaymentRailsERC20RequestV1
```
Comment (volodymyr-basiuk) [inline protocol/payment.go:75]: let's call it railsERC20

File: protocol/payment.go [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+	case Iden3PaymentRailsERC20V1Type:
+		return json.Marshal(p.railsERC)
+	}
+	return nil, errors.New("failed to marshal not initialized Payment")
+}
+
+// Iden3PaymentCryptoV1 represents the Iden3PaymentCryptoV1 payment data.
+type Iden3PaymentCryptoV1 struct {
+	ID          string         `json:"id"`
+	Type        string         `json:"type"`
```
Comment (volodymyr-basiuk) [inline protocol/payment.go:415]: Could we define separate type - iden3comm.PaymentType or similar ? 
For this file `Type` only one value are possible - `Iden3PaymentRequestCryptoV1`. 
The same for  Iden3PaymentRailsV1 and Iden3PaymentRailsERC20V1

**Review by vmidyllic — DISMISSED** [review]

File: protocol/payment.go (hunk 6 above)
Comment (x1m3) [reply protocol/payment.go]: Done.

**Review by volodymyr-basiuk — APPROVED** [review]

**Review by vmidyllic — APPROVED** [review]

**Pull request merged**